src/common/common.cpp

```diff
@@ -33,47 +33,99 @@ std::optional<POINT> get_mouse_pos() {
   }
 }
 
-HWND get_filtered_active_window() {
+
+
+static WindowAndProcPath get_filtered_base_window_and_path_impl(HWND window) {
   static auto desktop = GetDesktopWindow();
   static auto shell = GetShellWindow();
   static HWND searchui = nullptr;
-  auto active_window = GetForegroundWindow();
-  active_window = GetAncestor(active_window, GA_ROOT);
+  WindowAndProcPath rval;
+  window = GetAncestor(window, GA_ROOT);
 
-  if (active_window == desktop || active_window == shell || active_window == searchui) {
-    return nullptr;
+  if (window == nullptr || window == desktop || window == shell || window == searchui) {
+    return rval;
   }
-  
-  auto window_styles = GetWindowLong(active_window, GWL_STYLE);
+
+  auto window_styles = GetWindowLong(window, GWL_STYLE);
   if ((window_styles & WS_CHILD) || (window_styles & WS_DISABLED)) {
-    return nullptr;
+    return rval;
   }
   
-  window_styles = GetWindowLong(active_window, GWL_EXSTYLE);
+  window_styles = GetWindowLong(window, GWL_EXSTYLE);
   if ((window_styles & WS_EX_TOOLWINDOW) ||(window_styles & WS_EX_NOACTIVATE)) {
-    return nullptr;
+    return rval;
   }
   
   char class_name[256] = "";
-  GetClassNameA(active_window, class_name, 256);
+  GetClassNameA(window, class_name, 256);
   if (strcmp(class_name, "SysListView32") == 0 ||
       strcmp(class_name, "WorkerW") == 0 ||
       strcmp(class_name, "Shell_TrayWnd") == 0 ||
       strcmp(class_name, "Shell_SecondaryTrayWnd") == 0 ||
       strcmp(class_name, "Progman") == 0) {
-    return nullptr;
+    return rval;
   }
   if (strcmp(class_name, "Windows.UI.Core.CoreWindow") == 0) {
     const static std::wstring cortana_app = L"SearchUI.exe";
-    auto process_path = get_process_path(active_window);
+    auto process_path = get_process_path(window);
     if (process_path.length() >= cortana_app.length() &&
         process_path.compare(process_path.length() - cortana_app.length(), cortana_app.length(), cortana_app) == 0) {
       // cache the cortana HWND
-      searchui = active_window;
-      return nullptr;
+      searchui = window;
+      return rval;
     }
+    rval.process_path = std::move(process_path);
   }
-  return active_window;
+  if (rval.process_path.empty())
+  {
+      rval.process_path = get_process_path(window);
+  }
+  rval.hwnd = window;
+  return rval;
+}
+
+WindowAndProcPath get_filtered_base_window_and_path(HWND window) {
```

Comment:
I'm not sure I understand the way the caching is implemented, can you add a comment to give an overview of how it works?